src/image/filterRenderer2D.js

```diff
@@ -0,0 +1,207 @@
+   * @param {string} operation - The filter operation type (e.g., constants.BLUR).
+   * @param {string} filterParameter - The strength of applying filter.
+   * @param {p5.Shader} customShader - Optional custom shader; if provided, ignore operation-based loading.
+   */
+  constructor(pInst, operation, filterParameter, customShader) {
+    this.pInst = pInst;
+    this.filterParameter = filterParameter;
+    this.operation = operation;
+    this.customShader = customShader;
+
+
+    // Create a canvas for applying WebGL-based filters
+    this.canvas = document.createElement('canvas');
+    this.canvas.width = pInst.width;
+    this.canvas.height = pInst.height;
+
+    // Initialize the WebGL context
+    this.gl = this.canvas.getContext('webgl');
+    if (!this.gl) {
+      console.error("WebGL not supported, cannot apply filter.");
+      return;
+    }
+
+    // Minimal renderer object required by p5.Shader and p5.Texture
+    this._renderer = {
+      GL: this.gl,
+      registerEnabled: new Set(),
+      _curShader: null,
+      _emptyTexture: null,
+      webglVersion: 'WEBGL',
+      states: {
+        textureWrapX: this.gl.CLAMP_TO_EDGE,
+        textureWrapY: this.gl.CLAMP_TO_EDGE,
+      },
+      _arraysEqual: (a, b) => JSON.stringify(a) === JSON.stringify(b),
+      _getEmptyTexture: () => {
+        if (!this._emptyTexture) {
+          const im = new Image(1, 1);
+          im.set(0, 0, 255);
+          this._emptyTexture = new Texture(this._renderer, im);
+        }
+        return this._emptyTexture;
+      },
+    };
+
+    // Fragment shaders mapped to filter operations
+    this.filterShaders = {
+      [constants.BLUR]: filterBlurFrag,
+      [constants.INVERT]: filterInvertFrag,
+      [constants.THRESHOLD]: filterThresholdFrag,
+      [constants.ERODE]: filterErodeFrag,
+      [constants.GRAY]: filterGrayFrag,
+      [constants.DILATE]: filterDilateFrag,
+      [constants.POSTERIZE]: filterPosterizeFrag,
+      [constants.OPAQUE]: filterOpaqueFrag,
+    };
+
+    this._shader = null;
+    this._initializeShader();
+
+    // Create buffers once
+    this.vertexBuffer = this.gl.createBuffer();
+    this.texcoordBuffer = this.gl.createBuffer();
+
+    // Set up the vertices and texture coordinates for a full-screen quad
+    this.vertices = new Float32Array([-1, -1, 1, -1, -1, 1, 1, 1]);
+    this.texcoords = new Float32Array([0, 1, 1, 1, 0, 0, 1, 0]);
+
+    // Upload vertex data once
+    this._bindBufferData(this.vertexBuffer, this.gl.ARRAY_BUFFER, this.vertices);
+
+    // Upload texcoord data once
+    this._bindBufferData(this.texcoordBuffer, this.gl.ARRAY_BUFFER, this.texcoords);
+
+  }
+
+  updateFilterParameter(newFilterParameter) {
+    // Operation is the same, just update parameter if changed
+    this.filterParameter = newFilterParameter;
+  }
+
+  /**
+   * Initializes the shader program if it hasn't been already.
+   */
+  _initializeShader() {
+    if (this._shader) return; // Already initialized
```

Comment:
It looks like once we set `this._shader`, it will never update. I think that will cause this scenario to apply the first filter twice instead  of doing each one once:
```js
function draw() {
  filter(BLUR)
  filter(THRESHOLD)
}
```

I think maybe we should structure this in a way where we have `this.filterShaderSources` (storing the fragment shaders) and `this.filterShaders` (storing the initialized `p5.Shader`, if it exists). So in `_initializeShader`, we could just return the custom shader if a custom shader has been provided, and if not, return the stored `this.filterShaders[this.operation]` if it already exists, and if not, initialize a shader and store it in `this.filterShaders`.

Reply:
This would be instead of having separate filter renderers per shader as you have it right now. The browser has a pretty low limit for the number of active WebGL contexts it can maintain (it's something like 6, and lower on phones), so we'd want to have just one filter renderer per 2D context, and have it have multiple shaders.